# agb2025
Repository for the AGB 2025 common class project

[![DocumentationÂ â€¯Wiki](https://img.shields.io/static/v1?label=Documentation&message=Wiki&labelColor=black&color=blue&logo=github&logoColor=white)](https://github.com/egenomics/agb2025/wiki)


Wiki you can modify and use to document the process!

## Context (Here we will change it with the name of the pipeline)

This is a pipeline created in the Hospital del Mar bioinformaticians group (HdMBioinfo) to process microbiota data coming from stool samples processed in the hospital's analytical facilities by petitions of the Digestology Unit. The main objective is to classify the samples between healthy and non-healthy individuals, so the clinicians afterwards can decide useful interventions (ex: if that individual could be a potential stool donor, or if the individual needs from a stool transplant). The sample metadata comes from the hospital database, we retrieve it using the sampleID provided by the barcodes in the stool samples. Once the sample has been processed by a technician (mainly doing the DNA extraction), the sample goes into an Hi-Seq Illumina Sequencer. As we are performing advanced analysis of the samples, a small team made by bioinformatician technicians do the demultiplexing and send to us the FastQ files from each sample processed in that sequencing run.

## Project Directory Structure

```text
â”œâ”€â”€ ğŸ“ raw_data/            # Original FASTQ files (paired-end reads)
â”‚   â”œâ”€â”€ sample1_R1.fastq.gz
â”‚   â”œâ”€â”€ sample1_R2.fastq.gz
â”‚   â””â”€â”€ ...
â”œâ”€â”€ ğŸ“ metadata/            # Sample metadata (CSV or TSV)
â”‚   â””â”€â”€ run_metadata.csv
â”œâ”€â”€ ğŸ“ scripts/             # Custom or pipeline scripts
â”‚   â””â”€â”€ â€¦
â”œâ”€â”€ ğŸ“ outputs/              # Outputs per run
â”‚   â””â”€â”€ ğŸ“ run_YYYYmmdd_hhmmss
â”‚       â”œâ”€â”€ ğŸ“ qc_reports/          # FastQC or MultiQC reports
â”‚       â”‚   â”œâ”€â”€ fastqc/
â”‚       â”‚   â””â”€â”€ multiqc_report.html
â”‚       â”œâ”€â”€ ğŸ“ processed_data/      # Trimmed, filtered reads
â”‚       â”‚   â”œâ”€â”€ data_filtered/
â”‚       â”‚   â””â”€â”€ qiime2_demux/
â”‚       â”œâ”€â”€ ğŸ“ feature_tables/      # OTU or ASV tables
â”‚       â”‚   â”œâ”€â”€ table.qza
â”‚       â”‚   â””â”€â”€ table.tsv
â”‚       â”œâ”€â”€ ğŸ“ taxonomy/            # Taxonomy assignment results
â”‚       â”‚   â”œâ”€â”€ classifier.qza
â”‚       â”‚   â”œâ”€â”€ taxonomy.qza
â”‚       â”‚   â””â”€â”€ taxonomy.tsv
â”‚       â”œâ”€â”€ ğŸ“ phylogeny/           # Phylogenetic tree files
â”‚       â”‚   â”œâ”€â”€ rooted-tree.qza
â”‚       â”‚   â””â”€â”€ unrooted-tree.qza
â”‚       â”œâ”€â”€ ğŸ“ diversity_analysis/  # Alpha & beta diversity
â”‚       â”‚   â”œâ”€â”€ core-metrics-results/
â”‚       â”‚   â””â”€â”€ emperor_plots/
â”‚       â””â”€â”€ ğŸ“ visualizations/      # QIIME 2 visualizations
â”‚           â”œâ”€â”€ taxonomy_barplots.qzv
â”‚           â”œâ”€â”€ rarefaction_curves.png
â”‚           â””â”€â”€ â€¦
â”œâ”€â”€ ğŸ“ results/             # Final summary tables and plots per run (or sample?)
â”‚   â””â”€â”€ summary_report.html
â”‚ 
â””â”€â”€ ğŸ“ logs/                # Pipeline logs
    â””â”€â”€ run_YYYYmmdd_hhmmss_pipeline.log

